### ISSUE TITLE
Passt network binding plugin fails on Ubuntu Noble 24.04

### ISSUE THREAD

**Issue by bencoxford** [kubevirt/kubevirt#12333]
**What happened**:

Spin up a VM using the network binding plugin for passt interfaces following https://kubevirt.io/user-guide/network/net_binding_plugins/passt/ on Ubuntu 24.04 (Noble). The VM fails to start, citing the errors:

```
{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"vm-1","namespace":"vm","pos":"server.go:202","reason":"virError(Code=1, Domain=0, Message='internal error: Child process (/usr/bin/passt --one-off --socket /var/run/libvirt/qemu/run/passt/15-vm_vm-1-ua-default.socket --mac-addr 52:54:00:22:d9:3d --pid /var/run/libvirt/qemu/run/passt/15-vm_vm-1-ua-default-passt.pid --interface eth0 --log-file /var/run/kubevirt/passt.log --tcp-ports '~15000,~15001,~15004,~15006,~15008,~15009,~15020,~15021,~15053,~15090,3389,443') unexpected exit status 1: Couldn't create user namespace: Permission denied\n')","timestamp":"2024-07-12T10:25:51.627875Z","uid":"ea8dfc32-eda8-4300-9c86-0c6a201d09fb"}
```

The same configuration works on Ubuntu 22.04 (Jammy). It seems related to https://github.com/canonical/lxd/issues/12882:

```
k exec -ti -n vm virt-launcher-vm-1-786s6 -- unshare -U --mount-proc /bin/bash
unshare: unshare failed: Permission denied
command terminated with exit code 1
```

**What you expected to happen**:
 
VM should start with no errors.

**How to reproduce it (as minimally and precisely as possible)**:
Steps to reproduce the behavior.

Deploy a VM using passt network binding on Ubuntu (24.04) Noble. This requires multus to be installed.

**Additional context**:

https://github.com/canonical/lxd/issues/12882
https://github.com/canonical/lxd/issues/11920

**Environment**:
- KubeVirt version (use `virtctl version`): v1.2
- Kubernetes version (use `kubectl version`): v1.28.9
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: https://kubevirt.io/user-guide/network/net_binding_plugins/passt/
- OS (e.g. from /etc/os-release): Ubuntu 24.04 (Jammy)
- Kernel (e.g. `uname -a`): 6.8.0-38-generic
- Install tools: KubeVirt and Multus
- Others: N/A

**Comment by bc185174** [comment]
https://gitlab.com/apparmor/apparmor/-/wikis/unprivileged_userns_restriction

**Comment by sbrivio-rh** [comment]
@bc185174, thanks for reporting this.

_If_ passt's AppArmor profile (shipped with Ubuntu packages) is used in this case (do you know if it is?), I guess we need to add {{allow userns create}} to it, but we need to pay attention to avoid breaking Debian and openSUSE, and https://gitlab.com/apparmor/apparmor/-/wikis/unprivileged_userns_restriction#support-matrix is unfortunately Ubuntu-only.

@vasiliy-ul, do you happen to have some recommendation on how to approach this?

By the way, I just realised that [this issue(https://github.com/kubevirt/kubevirt/issues/4290#issuecomment-2149408179) is also still open. :(

**Comment by vasiliy-ul** [comment]
Well, I think this issue fits with my previous comment https://github.com/kubevirt/kubevirt/issues/11471#issuecomment-1991552664.

If it's the **_host_** profile that is blocking passt, then the recommendation from **_upstream_** KubeVirt perspective should be the same as written in the docs: https://kubevirt.io/user-guide/cluster_admin/installation/#integration-with-apparmor, i.e. remove the respective apparmor profiles from the host.

Alternatively, if it is the default profile of the container runtime that is causing the denial, then it is similar to the case with virtiofsd and the mount call. Unfortunately, here we have only workarounds at the moment: either run unconfined or use the suggested sysctl.

**Comment by bc185174** [comment]
Thank you, I'll give installing the `passt` deb pkg and see how it goes.

>  i.e. remove the respective apparmor profiles from the host.

I don't see anywhere in the docs that this is the approach KubeVirt recommends in terms of app armor profiles?

**Comment by vasiliy-ul** [comment]
>I don't see anywhere in the docs that this is the approach KubeVirt recommends in terms of app armor profiles?

It is mentioned in the user guide: https://kubevirt.io/user-guide/cluster_admin/installation/#integration-with-apparmor

> The first (and simplest) one is to remove the libvirt package from the host: assuming the host is a dedicated Kubernetes node, you likely won't need it anyway.

The removal of libvirt pckage will also remove the respective profiles from the host.

**Comment by sbrivio-rh** [comment]
> If it's the **_host_** profile that is blocking passt, then the recommendation from **_upstream_** KubeVirt perspective should be the same as written in the docs: https://kubevirt.io/user-guide/cluster_admin/installation/#integration-with-apparmor, i.e. remove the respective apparmor profiles from the host.

From my perspective (as passt developer and maintainer of the Debian package)... I would like to fix this, if there's something I can fix. So I would need these two bits of information:
- is it really the host profile blocking passt?
- what's the recommendation around that `allow userns create` for openSUSE? I don't want to risk "fixing" the profile for Debian/Ubuntu and breaking it for openSUSE...

**Comment by bc185174** [comment]
Just installed passt app armor profiles and it worked.

**Comment by sbrivio-rh** [comment]
> Just installed passt app armor profiles on host and it worked.

Thanks for testing. Weird, they don't include any rule to explicitly permit user namespace creation. Perhaps it's because they use ABI version 3.0 (https://gitlab.com/apparmor/apparmor/-/wikis/unprivileged_userns_restriction#apparmor-3x) and therefore `allow userns create` is assumed by default?

Well, at least that's something I can try out on Ubuntu 24.04 separately from KubeVirt.

Still, I wonder if it's correct that the behaviour of passt in KubeVirt depends on whether the package is installed on the host. In particular, if the package is not installed on the host, I guess the AppArmor profile applied (if any) is not as strict as the one from the passt package, because the former is not tailored around passt.

**Comment by sbrivio-rh** [comment]
I finally had the chance to check the behaviour of stand-alone passt on Ubuntu 23.10, 24.04, a current snapshot of the upcoming 24.10, a current openSUSE Tumbleweed version, and a current Debian unstable (sid) installation.

As long as the AppArmor profiles shipped with the packages are present, there are no issues. The AppArmor ABI of the profile is `3.0`, so it doesn't contain an explicit `allow userns create`, but the mere fact that there's a profile with ABI 3.0 allows pasta to create its sandboxing user namespace.

Now, while Ubuntu 24.10 and openSUSE Tumbleweed ship AppArmor packages with support for the `4.0` ABI, Debian unstable still ships 3.1.17, so, to keep things simple and still ship a single AppArmor profile (developed upstream), I won't update the profile to ABI 4.0 yet. Updating the profile wouldn't solve the issue with KubeVirt anyway.

I still believe KubeVirt needs proper AppArmor integration, so I don't think that this issue should be closed. For example, in this case, `passt` could be required as a dependency on the host KubeVirt is running on. Or profiles need to be shipped, somehow.

But running things without Linux Security Modules profiles they were supposed to run with is in my opinion an **important security weakness** which needs to be addressed.

**Comment by oshoval** [comment]
@EdDev
Hi, might interest you to know this about passt
Thanks

**Comment by bc185174** [comment]
> Now, while Ubuntu 24.10 and openSUSE Tumbleweed ship AppArmor packages with support for the `4.0` ABI, Debian unstable still ships 3.1.17, so, to keep things simple and still ship a single AppArmor profile (developed upstream), I won't update the profile to ABI 4.0 yet. Updating the profile wouldn't solve the issue with KubeVirt anyway.

Yeah ignore my last comment I had used `sysctl` to allow it so even though installed `passt` deb pkg still hitting same issue. You're right it won't work with the profile installed on host. 

> I still believe KubeVirt needs proper AppArmor integration, so I don't think that this issue should be closed. For example, in this case, passt could be required as a dependency on the host KubeVirt is running on. Or profiles need to be shipped, somehow.
> But running things without Linux Security Modules profiles they were supposed to run with is in my opinion an important security weakness which needs to be addressed.

For sure need proper AppArmor integration.

**Comment by trevex** [comment]
I wanted to test passt and ran into the same issue. I am running a Custer on Talos however. From my understanding AppArmor support was only recently introduced to Talos with 1.8.0 and I am still running 1.7.X.

From the logs of the compute container of VMs using passt, I can see that passt fails with:
```
{"component":"virt-launcher","level":"error","msg":"internal error: Child process (passt --one-off --socket /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket --pid /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default-passt.pid --interface eth0 --log-file /var/run/kubevirt/passt.log --tcp-ports all --udp-ports all) unexpected exit status 1: UNIX domain socket bound at /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket","pos":"virCommandWait:2773","subcomponent":"libvirt","thread":"33","timestamp":"2024-10-01T20:20:49.254000Z"}
{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to start VirtualMachineInstance with flags 0.","name":"tenant-a-cluster-worker-7x5k2-8p4kf","namespace":"tenant-a","pos":"manager.go:1251","reason":"virError(Code=1, Domain=0, Message='internal error: Child process (passt --one-off --socket /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket --pid /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default-passt.pid --interface eth0 --log-file /var/run/kubevirt/passt.log --tcp-ports all --udp-ports all) unexpected exit status 1: UNIX domain socket bound at /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket\nCouldn't create user namespace: Operation not permitted\n')","timestamp":"2024-10-01T20:20:49.255822Z","uid":"66b68f5e-f079-437b-ae54-041f9a0d7b72"}
{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"tenant-a-cluster-worker-7x5k2-8p4kf","namespace":"tenant-a","pos":"server.go:202","reason":"virError(Code=1, Domain=0, Message='internal error: Child process (passt --one-off --socket /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket --pid /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default-passt.pid --interface eth0 --log-file /var/run/kubevirt/passt.log --tcp-ports all --udp-ports all) unexpected exit status 1: UNIX domain socket bound at /var/run/libvirt/qemu/run/passt/18-tenant-a_tenant-a-cl-ua-default.socket\nCouldn't create user namespace: Operation not permitted\n')","timestamp":"2024-10-01T20:20:49.255862Z","uid":"66b68f5e-f079-437b-ae54-041f9a0d7b72"}
```

**Comment by sbrivio-rh** [comment]
> From my understanding AppArmor support was only recently introduced to Talos with 1.8.0 and I am still running 1.7.X.

If your understanding is correct, then it must be something else... `cat /proc/sys/kernel/unprivileged_userns_clone`?

**Comment by trevex** [comment]
`/proc/sys/kernel/unprivileged_userns_clone` does not exist, but to my surprise `/proc/sys/kernel/unprivileged_userns_apparmor_policy` and it is set to `1`. Let me try the to disable it and test again.

**Comment by trevex** [comment]
The issue on Talos was not related to AppArmor, but rather that on Talos the Seccomp RuntimeDefault Profile is enforced. Explicitly setting a profile resolves this issue. Related: https://github.com/kubevirt/kubevirt/issues/10528

**Comment by sbrivio-rh** [comment]
> The issue on Talos was not related to AppArmor, but rather that on Talos the Seccomp RuntimeDefault Profile is enforced. Explicitly setting a profile resolves this issue. Related: #10528

@trevex, thanks for following up. I suppose the problematic part of the seccomp profile is not allowing unshare(2) with `CLONE_NEWUSER | CLONE_NEWNET` as `flags`, correct?

Given that #10528 is closed, I'm trying to understand if 1. there's anything we can do in passt (but I don't think so, we need that for sandboxing and we run with an otherwise much more restrictive profile...), 2. what we should do in KubeVirt (if anything). The RuntimeDefault profile that's applied on Talos comes from... KubeVirt itself? Docker?

**Comment by trevex** [comment]
@sbrivio-rh Yes basically containerd, it has been some weeks, but if I remember correctly, Talos uses the RuntimeDefault seccomp profile even if it is not specified by pods. 

So I don't think this is something KubeVirt has to address.

**Comment by bc185174** [comment]
I've just revisited this and integrated it with app armor. If you are running k8s v1.31+ you can set the following app armour profile on the `virt-launcher` pod. Its based on the containerd default but gives it access to `mount` and `userns`. It also solves the known issue https://github.com/kubevirt/kubevirt/issues/4290.

I'd suggest adding a similar feature gate `KubevirtAppArmorProfile` to write the app armor profile to `/etc/apparmor.d/kubevirt.vm` and modify the security context on the `virt-launcher` pod to use this profile.

```
abi <abi/3.0>,

include <tunables/global>

profile kubevirt.passt.vm flags=(attach_disconnected) {
  include <abstractions/passt>

  # Alternatively: include <abstractions/user-tmp>
  owner /tmp/**                         w,      # tap_sock_unix_init(), pcap(),
                                                # write_pidfile(),
                                                # logfile_init()

  owner @{HOME}/**                      w,      # pcap(), write_pidfile()
  mount,
  userns,

  network,
  capability,
  file,
  umount,
  # Host (privileged) processes may send signals to container processes.
  signal (receive) peer=unconfined,
  # runc may send signals to container processes.
  signal (receive) peer=runc,
  # crun may send signals to container processes.
  signal (receive) peer=crun,

  deny @{PROC}/* w,   # deny write for all files directly in /proc (not in a subdir)
  deny @{PROC}/{[^1-9],[^1-9][^0-9],[^1-9s][^0-9y][^0-9s],[^1-9][^0-9][^0-9][^0-9]*}/** w,
  deny @{PROC}/sys/[^k]** w,  # deny /proc/sys except /proc/sys/k* (effectively /proc/sys/kernel)
  deny @{PROC}/sys/kernel/{?,??,[^s][^h][^m]**} w,  # deny everything except shm* in /proc/sys/kernel/
  deny @{PROC}/sysrq-trigger rwklx,
  deny @{PROC}/mem rwklx,
  deny @{PROC}/kmem rwklx,
  deny @{PROC}/kcore rwklx,

  deny /sys/[^f]*/** wklx,
  deny /sys/f[^s]*/** wklx,
  deny /sys/fs/[^c]*/** wklx,
  deny /sys/fs/c[^g]*/** wklx,
  deny /sys/fs/cg[^r]*/** wklx,
  deny /sys/firmware/** rwklx,
  deny /sys/devices/virtual/powercap/** rwklx,
  deny /sys/kernel/security/** rwklx,
}
```

**Comment by sbrivio-rh** [comment]
> Rotten issues close after 30d of inactivity.

Why? This hasn't been solved yet, so it shouldn't be closed.

/reopen

**Comment by bencoxford** [comment]
We've got past this for now by setting the seccomp/apparmor profile on kubevirt compute containers.

```
pod.Spec.Containers[idx].SecurityContext.SeccompProfile = &corev1.SeccompProfile{
	Type:             corev1.SeccompProfileTypeLocalhost,
	LocalhostProfile: &kubeVirtSeccompProfile,
}
pod.Spec.Containers[idx].SecurityContext.AppArmorProfile = &corev1.AppArmorProfile{
	Type:             corev1.AppArmorProfileTypeLocalhost,
	LocalhostProfile: &kubeVirtAppArmorProfile,
}
```

Can we add some configuration to set this? Using webhook injection is not ideal as `virt-controller` doesn't like us modifying the pod spec. The other proposal is to let the `virt-controller` create a managed `Deployment` instead of `Pod` so we can use mutating webhooks.

@sbrivio-rh @vasiliy-ul ^

**Comment by nirdothan** [comment]
Thanks @bc185174! That's great news. 
@xpivarc do you think that it would be possible to introduce  AppArmor profiles into the API in a similar manner to [seccomp profile](https://github.com/kubevirt/kubevirt/blob/release-1.5/staging/src/kubevirt.io/api/core/v1/types.go#L2561)?
